src/pdf.sandbox.js

```diff
@@ -24,86 +25,124 @@ const TESTING =
   typeof PDFJSDev === "undefined" || PDFJSDev.test("!PRODUCTION || TESTING");
 
 class Sandbox {
-  constructor(module) {
-    this._evalInSandbox = module.cwrap("evalInSandbox", null, [
-      "string",
-      "int",
-    ]);
-    this._dispatchEventName = null;
+  constructor(module, externalNames) {
+    this._commFun = module.cwrap("commFun", null, ["string", "string"]);
     this._module = module;
-    this._alertOnError = 1;
+    this._alertOnError = 0;
+    this._externalNames = externalNames;
   }
 
   create(data) {
-    const sandboxData = JSON.stringify(data);
-    const extra = [
-      "send",
-      "setTimeout",
-      "clearTimeout",
-      "setInterval",
-      "clearInterval",
-      "crackURL",
-    ];
-    const extraStr = extra.join(",");
-    let code = [
-      "exports = Object.create(null);",
-      "module = Object.create(null);",
+    const sandboxData = Sandbox.cloneIn(data);
+    // In the sandox, external functions like send, setTimeout, ... exist
+    // as global variables.
+    const external = `{ ${this._externalNames.join(",")} }`;
+    const code = [
       // Next line is replaced by code from initialization.js
       // when we create the bundle for the sandbox.
       PDFJSDev.eval("PDF_SCRIPTING_JS_SOURCE"),
-      `data = ${sandboxData};`,
-      `module.exports.initSandbox({ data, extra: {${extraStr}}, out: this});`,
-      "delete exports;",
-      "delete module;",
-      "delete data;",
+      `pdfjsScripting.initSandbox({ data: ${sandboxData}, external: ${external} })`,
     ];
     if (!TESTING) {
-      code = code.concat(extra.map(name => `delete ${name};`));
       code.push("delete debugMe;");
+      // Don't expose external functions
+      code.push(this._externalNames.map(name => `delete ${name};`).join(";"));
+      delete this._externalNames;
```

Comment:
This line doesn't do anything bad, but it's not doing anything useful either, except potentially preventing this `create` sandbox from being called twice (because a second call would trigger a runtime error when `.join()` is called). Is that what you've meant with adding this code? With all `delete` and so on around here, I just want to double-check that the code matches with your intentions.